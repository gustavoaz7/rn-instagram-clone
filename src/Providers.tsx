import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import { Provider } from 'react-redux';
import { theme as stylesTheme } from './styles/theme';
import { store as reduxStore } from './redux/store';

type TProvidersProps = {
  children: ReactNode;
  store?: typeof reduxStore;
  theme?: typeof stylesTheme;
};

export function Providers({
  children,
  store = reduxStore,
  theme = stylesTheme,
}: TProvidersProps): JSX.Element {
  return (
    <Provider store={store}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </Provider>
  );
}
