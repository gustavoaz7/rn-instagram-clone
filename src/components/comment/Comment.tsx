import React, { memo } from 'react';
import styled from 'styled-components/native';
import { Text } from '../text';
import { AvatarWithRing } from '../avatar-with-ring';
import HeartSvg from '../../../assets/svg/heart.svg';
import { TComment } from '../../types';
import { pluralizeWithS } from '../../utils/string';
import { dateToString } from '../../utils/date';

export const AVATAR_SIZE = 40;

type TCommentProps = TComment;

export const Comment = memo(function Comment(
  comment: TCommentProps,
): JSX.Element {
  const likeCount = comment.likedBy.length;
  return (
    <Container>
      <AvatarWithRing
        size={AVATAR_SIZE}
        color="gradient"
        imageUrl={comment.owner.profilePicUrl}
      />
      <CommentContainer>
        <CommentText>
          <BoldText>{comment.owner.username} </BoldText>
          {comment.text}
        </CommentText>
        <Row>
          <CommentStats>
            {dateToString(new Date(comment.createdAt), true)}
          </CommentStats>
          {likeCount > 0 ? (
            <CommentStats>
              {likeCount} {pluralizeWithS('like', likeCount)}
            </CommentStats>
          ) : null}
          <CommentStats>Reply</CommentStats>
        </Row>
      </CommentContainer>
      <HeartIcon />
    </Container>
  );
});

const Row = styled.View`
  flex-direction: row;
`;

const BoldText = styled(Text)`
  font-weight: bold;
`;

const CommentStats = styled(BoldText)`
  color: ${({ theme }) => theme.color.gray};
  margin-right: ${({ theme }) => theme.spacing.m};
  font-size: ${({ theme }) => theme.font.size.s};
`;

const Container = styled(Row)`
  justify-content: space-between;
  padding: 0 ${({ theme }) => theme.spacing.l};
`;

const CommentContainer = styled.View`
  flex: 1;
  padding-left: ${({ theme }) => theme.spacing.m};
`;

const CommentText = styled(Text)`
  margin-right: ${({ theme }) => theme.spacing.xs};
  margin-bottom: ${({ theme }) => theme.spacing.xs};
`;

const HeartIcon = styled(HeartSvg).attrs(({ theme }) => ({
  color: theme.color.gray,
  width: 16,
  height: 16,
}))`
  margin-top: ${({ height }) => AVATAR_SIZE / 2 - height / 2}px;
`;
