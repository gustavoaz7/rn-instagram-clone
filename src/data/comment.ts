import { TComment } from '../types';

export const generateComment = (): TComment => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
    username: 'Vivienne_Doyle',
  },
  createdAt: 1620488304326,
  text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
  likedBy: ['Frankie30', 'Patricia24', 'Westley_Ebert49'],
});
