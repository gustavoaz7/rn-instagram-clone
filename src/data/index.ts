import type {
  TLike,
  TComment,
  TPostMedia,
  TPost,
  TOwner,
  TStory,
  TStoryMedia,
  TProfile,
  TTappableObject,
} from '../types';

export const generateMockOwner = (): TOwner => ({
  id: `${Math.random()}`,
  profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
  username: 'gustavoaz7',
});

export const generateMockLike = (): TLike => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  createdAt: 1620488304326,
});

export const generateMockComment = (likesQty = 1): TComment => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
    username: 'Vivienne_Doyle',
  },
  createdAt: 1620488304326,
  text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
  previewLikes: {
    count: likesQty,
    likes: [...Array(likesQty)].map(generateMockLike),
  },
  viewerHasLiked: false,
});

export const generateMockTappableObject = ({
  type = 'mention',
  text = 'random',
  x = 0.1,
  y = 0.2,
  scale = 1.1,
  rotation = 300,
}: Partial<TTappableObject> = {}): TTappableObject => ({
  id: `${Math.random()}`,
  type,
  text,
  x,
  y,
  scale,
  rotation,
});

export const generateMockPostMedia = ({ index = 0 } = {}): TPostMedia => ({
  id: `${Math.random()}`,
  url: `http://placeimg.com/640/480/cats?${index}`,
  owner: generateMockOwner(),
  tappableObjects: [],
});

export const generateMockPost = ({
  mediasQty = 1,
  likesQty = 3,
  commentsQty = 1,
} = {}): TPost => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  createdAt: 1620488304326,
  medias: [...Array(mediasQty)].map(generateMockPostMedia),
  caption:
    'Tempora rerum minima dolor quibusdam blanditiis ea et. Earum iste iure expedita velit laudantium inventore quia qui. Distinctio consectetur accusamus voluptatum hic rem at quas. Incidunt ipsum sed illo molestiae vitae ullam sit consequatur autem. Ut voluptatem nisi. Velit ex dignissimos sed aut.',
  previewLikes: {
    count: likesQty,
    likes: [...Array(likesQty)].map(generateMockLike),
  },
  previewComments: {
    count: commentsQty,
    comments: [...Array(commentsQty)].map(generateMockComment),
  },
  location: 'Daronville',
  viewerHasLiked: false,
});

export const generateMockStoryMedia = ({
  viewersQty = 0,
  index = 0,
} = {}): TStoryMedia => ({
  id: `${Math.random()}`,
  url: `http://placeimg.com/640/480/cats?${index}`,
  owner: generateMockOwner(),
  tappableObjects: [],
  previewViewers: {
    count: viewersQty,
    viewers: [...Array(viewersQty)].map(generateMockOwner),
  },
  takenAt: 1620488304326,
  expiresAt: 1620488304326 + 3600 * 24,
});

export const generateMockStory = ({
  mediasQty = 3,
  viewersQty = 0,
} = {}): TStory => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  expiresAt: 0,
  latestMediaAt: 0,
  seenAt: 0,
  medias: [...Array(mediasQty)].map((_, index) =>
    generateMockStoryMedia({ viewersQty, index }),
  ),
});

export const generateMockProfile = (): TProfile => ({
  bio: 'animi eum earum aut explicabo',
  followCount: 47,
  followedByCount: 50,
  followedByViewer: false,
  followsViewer: false,
  fullName: 'Gustavo Az',
  postsCount: 20,
  profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
  username: 'gustavoaz7',
  story: generateMockStory(),
});
