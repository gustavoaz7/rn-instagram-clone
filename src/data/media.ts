import type { TPostMedia } from '../types';

export const generatePostMedia = (): TPostMedia => ({
  id: `${Math.random()}`,
  url: 'http://placeimg.com/640/480/cats',
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
    username: 'gustavoaz7',
  },
  tappableObjects: [],
});
