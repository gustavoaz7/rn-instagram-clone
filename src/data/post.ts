import { TPost } from '../types';
import { generatePostMedia } from './media';

export const createMockPost = (): TPost => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
    username: 'gustavoaz7',
  },
  createdAt: 1620488304326,
  medias: [generatePostMedia()],
  caption:
    'Tempora rerum minima dolor quibusdam blanditiis ea et. Earum iste iure expedita velit laudantium inventore quia qui. Distinctio consectetur accusamus voluptatum hic rem at quas. Incidunt ipsum sed illo molestiae vitae ullam sit consequatur autem. Ut voluptatem nisi. Velit ex dignissimos sed aut.',
  likedBy: ['Tito.Zieme', 'Cleo.Runte69', 'Laverna.Emmerich24'],
  comments: [
    {
      id: `${Math.random()}`,
      owner: {
        id: `${Math.random()}`,
        profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
        username: 'Vivienne_Doyle',
      },
      createdAt: 1620488304326,
      text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
      likedBy: ['Frankie30', 'Patricia24', 'Westley_Ebert49'],
    },
  ],
  location: 'Daronville',
});
