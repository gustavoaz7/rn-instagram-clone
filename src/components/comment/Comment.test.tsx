import React from 'react';
import { render } from '@testing-library/react-native';
import { createMockPost } from '../../data/post';
import { Comment } from './Comment';
import { Providers } from '../../Providers';

describe('components - Comment', () => {
  const options = { wrapper: Providers };
  const comment = createMockPost().comments[0];

  it('renders', () => {
    render(<Comment {...comment} />, options);
  });

  it('matches snapshot', () => {
    const realDateNow = Date.now.bind(global.Date);
    const dateNowMock = jest.fn(() => comment.createdAt + 3600 * 1000);
    global.Date.now = dateNowMock;

    const { toJSON } = render(<Comment {...comment} />, options);

    expect(toJSON()).toMatchSnapshot();
    global.Date.now = realDateNow;
  });

  describe('likes', () => {
    it('renders without likes', () => {
      render(<Comment {...comment} likedBy={[]} />, options);
    });

    it('renders singular when 1 like', () => {
      const { getByText } = render(
        <Comment {...comment} likedBy={['single']} />,
        options,
      );

      expect(getByText('1 like')).toBeTruthy();
    });

    it('renders plural when > 2 likes', () => {
      const { getByText } = render(
        <Comment {...comment} likedBy={['1', '2']} />,
        options,
      );

      expect(getByText('2 likes')).toBeTruthy();
    });
  });
});
