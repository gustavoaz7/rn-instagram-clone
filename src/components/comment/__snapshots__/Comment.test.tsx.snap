// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - Comment matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 0,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 0,
      },
    ]
  }
>
  <View
    size={40}
    style={
      Array [
        Object {
          "borderRadius": 40,
          "height": 40,
          "overflow": "hidden",
          "position": "relative",
          "width": 40,
        },
      ]
    }
  >
    <ViewManagerAdapter_ExpoLinearGradient
      proxiedProperties={
        Object {
          "colors": Array [
            4292948076,
            4294750021,
          ],
          "endPoint": Array [
            0.3,
            0.7,
          ],
          "locations": undefined,
          "startPoint": Array [
            0.7,
            0.3,
          ],
        }
      }
      style={
        Array [
          Object {
            "alignItems": "center",
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    >
      <View
        ringWidth={2}
        size={40}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "borderRadius": 36,
              "height": 36,
              "justifyContent": "center",
              "width": 36,
            },
          ]
        }
      >
        <Image
          offset={2}
          ringWidth={2}
          size={40}
          source={
            Object {
              "uri": "https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg",
            }
          }
          style={
            Array [
              Object {
                "borderColor": "#26262633",
                "borderRadius": 32,
                "borderWidth": 2,
                "height": 32,
                "width": 32,
              },
            ]
          }
        />
      </View>
    </ViewManagerAdapter_ExpoLinearGradient>
  </View>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingLeft": 15,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#262626",
          },
          Object {
            "marginBottom": 5,
            "marginRight": 5,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
            Object {
              "fontWeight": "bold",
            },
          ]
        }
      >
        Vivienne_Doyle
         
      </Text>
      Aspernatur quod ea unde repellendus magni nihil sunt.
    </Text>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
            Object {
              "color": "#8e8e8e",
              "fontSize": 12,
              "fontWeight": "bold",
              "marginRight": 15,
            },
          ]
        }
      >
        1 h
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
            Object {
              "color": "#8e8e8e",
              "fontSize": 12,
              "fontWeight": "bold",
              "marginRight": 15,
            },
          ]
        }
      >
        3
         
        likes
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
            Object {
              "color": "#8e8e8e",
              "fontSize": 12,
              "fontWeight": "bold",
              "marginRight": 15,
            },
          ]
        }
      >
        Reply
      </Text>
    </View>
  </View>
  <SvgMock
    color="#8e8e8e"
    height={16}
    style={
      Array [
        Object {
          "marginTop": 12,
        },
      ]
    }
    width={16}
  />
</View>
`;
